esmvaltool/diag_scripts/mlr/custom_sklearn.py

```diff
@@ -0,0 +1,771 @@
+        if return_var or return_cov:
+            pred = prediction[0]
+        else:
+            pred = prediction
+        if pred.ndim == 1:
+            pred_trans = self.transformer_.inverse_transform(
+                pred.reshape(-1, 1))
+        else:
+            pred_trans = self.transformer_.inverse_transform(pred)
+        if self._to_be_squeezed(pred_trans, always_return_1d=always_return_1d):
+            pred_trans = pred_trans.squeeze(axis=1)
+        if not (return_var or return_cov):
+            return pred_trans
+
+        # Return scaled variance or covariance if desired
+        err = prediction[1]
+        if not hasattr(self.transformer_, 'scale_'):
+            raise NotImplementedError(
+                f"Transforming of additional prediction output (e.g. by "
+                f"'return_var' or 'return_cov') is not supported for "
+                f"transformer {self.transformer_.__class__} yet, the "
+                f"necessary attribute 'scale_' is missing")
+        scale = self.transformer_.scale_
+        if scale is not None:
+            err *= scale**2
+        if self._to_be_squeezed(err, always_return_1d=always_return_1d):
+            err = err.squeeze(axis=1)
+        return (pred_trans, err)
+
+    def _get_fit_params(self, fit_kwargs):
+        """Separate ``transformer`` and ``regressor`` kwargs."""
+        steps = [
+            ('transformer', self.transformer),
+            ('regressor', self.regressor),
+        ]
+        fit_params = _get_fit_parameters(fit_kwargs, steps, self.__class__)
+
+        # FIXME
+        if fit_params['transformer']:
+            raise NotImplementedError(
+                f"Fit parameters {fit_params['transformer']} for transformer "
+                f"{self.transformer.__class__} of {self.__class__} are not "
+                f"supported at the moment")
+
+        return (fit_params['transformer'], fit_params['regressor'])
+
+    def _to_be_squeezed(self, array, always_return_1d=True):
+        """Check if ``array`` should be squeezed or not."""
+        squeeze = array.ndim == 2 and array.shape[1] == 1
+        if not always_return_1d:
+            squeeze = squeeze and self._training_dim == 1
+        return squeeze
+
+
+class FeatureSelectionTransformer(BaseEstimator, SelectorMixin):
+    """Transformer step of a feature selection estimator."""
+
+    def __init__(self, grid_scores, n_features, ranking, support):
+        """Initialize feature selection transformer."""
+        self.grid_scores = grid_scores
+        self.n_features = n_features
+        self.ranking = ranking
+        self.support = support
+
+    def fit(self, *_, **__):
+        """Empty method."""
+        return self
+
+    def _get_support_mask(self):
+        """Get support mask."""
+        return self.support
+
+    def _more_tags(self):
+        """Additional estimator tags."""
+        more_tags = super()._more_tags()
+        more_tags['allow_nan'] = True
+        return more_tags
```

Comment:
probably the best way to review this, rather than a line by line review, which is tedious and error-prone for everyone but you and the sklearn module devvs would be to summarize in a few bullet points what custom changes you made to the scipy module and add it to the documentation (something like technical stuff, if it's in your paper just lift it from there)

Reply:
I don't think further doc is necessary. The changes made are either in the name of the function, e.g. `_score_weighted` or `_fit_and_score_weighted` instead of `_score` and `_fit_and_score` or listed in the docstring, e.g. `"""Expand :meth:`fit` to accept kwargs."""` or `"""Expand :meth:`_fit` to accept kwargs."""`. Since this is rather technical stuff this is not in the paper :grimacing:

Reply:
okay - have you thought about contributing directly to `scipy` and see if these things can be included? I am a bit apprehensive to have a lot of code lifted from a library that evolves with time and dropped+changed locally (wrt `scipy`, not locally to us) - I am 100% sure of what you did is right but I don't want to have a wise guy come over to us in a year or so and "find bugs" and say we made a Chinese copy. If these changes were documented to the very least in a document we'd be able to say look we made these changes, tested them and it fit our purpose

Reply:
I fully agree with you - it would be optimal to have this in the `sklearn` itself, but I really do not have the time to do this right now, sorry :disappointed: I can expand the tests (there are already quite some) in a future PR to ensure that these custom changes are working if necessary, but I cannot provide any more documentation on this than already given by the docstring and the comments in the code, and moving them to a different document is not useful in my opinion.

Reply:
no, of course not now, that process is very lengthy - ok meet me middle ground then, how about a few lines worth of paragraph in a footnote in the documentation saying that you took these funcs and expanded them to fit the purposes for the desired functionality and you are open to contribute these features to the scipy's library - and mention that tests cover these as well and will increase the test coverage in the future even more. This should cover our arses and you don't have to do much more work too (well, apart from improving the test coverage in a future PR, after you finish writing your thesis :grin: )

Reply:
Agreed :+1: I added a note to documentation of the module, please check if you agree with it!